### ISSUE TITLE
Flag 'instanceof' expressions that are provably always true or false

### ISSUE THREAD

**Issue by erictraut** [microsoft/TypeScript#32801]
## Suggestion
When refactoring code, TypeScript is generally very good at finding and reporting situations where existing code will break due to type changes. This allows "refactoring with confidence". This isn't surprising given that one of the primary goals of TypeScript is to "statically identify constructs that are likely to be errors".

There's one common case where TypeScript is "blind" and doesn't report errors when types are refactored. This case involves the use of the 'instanceof' operator. This operator can be used in situations where it makes no sense -- e.g. where the type specified on the LHS has no possible relation to the type specified on the RHS.

To help developers avoid this common programming error, I propose that TypeScript report an error for any 'instanceof' expression that is provably always true or false at compilation time. Any such expression will likely be unintended by the programmer and should be flagged as errors. At best, such operations represent unnecessary code that imposes runtime overhead.

## Example

```typescript
// ClassB is a subclass of ClassA.
class ClassA {}
class ClassB extends ClassA {}

// ClassC has nothing to do with either
// ClassA or ClassB.
class ClassC {
    method1() {}
}

function function1(param1: ClassB) {
    // Static analysis can prove this expression
    // will always evaluate to false. It's a common
    // source of programming errors, especially during
    // refactoring.
    if (param1 instanceof ClassC) {
        // This call isn't even valid given that param1
        // was declared as type ClassB.
        param1.method1();
    }

    // Static analysis can prove this expression will
    // always evaluate to true, so at best it's unnecessary
    // runtime overhead and more commonly is a behavior
    // that is unintended by the programmer.
    if (param1 instanceof ClassA) {
        // ...
    }
}
```

Are there any legitimate uses of 'instanceof' where the result is statically provable to be true or false in all cases? I can't think of any, but it's possible I'm overlooking some specialized cases. If there are, this check could be added as an optional compiler switch to preserve the current behavior.

**Comment by erictraut** [comment]
Incidentally, I just added a similar feature (a "reportUnnecessaryIsInstance" switch) to [pyright](https://github.com/microsoft/pyright/releases/tag/1.0.55). Pyright is a Python type checker that borrows many concepts from TypeScript's type checker. Python's "isinstance" call is analogous to Javascript's "instanceof" operator. When I added this new feature to pyright, it immediately uncovered about a dozen bugs in my team's Python code base.

**Comment by stringham** [comment]
I've fixed a couple of bugs recently that would have been prevented if TypeScript would make this a compile error. After someone did a refactor they missed an instanceof check later in the function that now always returns false.

I think this would be similar to how this is an error:

```typescript
function foo(value: string) {
    return value === 'hello' && value === 'goodbye';
}
```

The compiler gives you this error: `This condition will always return 'false' since the types '"hello"' and '"goodbye"' have no overlap. (2367)`

**Comment by zen0wu** [comment]
+1. Just got bitten by this and I was assuming TS's checking for this and would prevent me from doing such silly things. I was trying to do
```
if (treeNode instanceof TreeValue) { ... } // TreeValue = typeof treeNode.value
```

In this case it should be trivial for TS to flag this is always false.

**Comment by paul-sachs** [comment]
+1 as well, just got burned by this.

**Comment by hazae41** [comment]
Days since last bug caused by this: 0
